### PR TITLE
Add hipchat and ordering support to redoctober

### PR DESCRIPTION
**Pull request by ejcx**
Supports MSP and requires several arguments to add hipchat integration to
red october. RedOctober will then alert on creation of an order, any new
delegation, or several other states.

### CODE REVIEW

File: core/core.go [hunk 1]
```diff
@@ -9,9 +9,13 @@ import (
 	"errors"
 	"fmt"
 	"log"
+	"strconv"
+	"time"
 
 	"github.com/cloudflare/redoctober/cryptor"
+	"github.com/cloudflare/redoctober/hipchat"
```
Comment (andrewbuss) [inline core/core.go]: This path doesn't exist. Should it have been committed too?

File: core/core.go [hunk 2]
```diff
@@ -658,3 +728,128 @@ func Export(jsonIn []byte) ([]byte, error) {
...
+		expiryTime,
+		duration,
+		adminsDelegated,
+		owners,
+		numDelegated)
+	orders.Orders[orderNum] = ord
+	out, err = json.Marshal(ord)
+
+	// We need to get the admins hipchat names.
+	hipchatOwners := make(map[string]string)
```
Comment (andrewbuss) [inline core/core.go]: I don't think this part of the code should care (or know) whether we're using Hipchat or another notification method. It'd be good IMO to refactor NotifyNewOrder to take only a list of usernames and perform this translation internally.

**Comment by andrewbuss** [comment]
The ordering system looks pretty good, but as written it seems like it won't work without Hipchat, when the orders functionality could stand on its own without notifications (i.e. just the API endpoints). It's good that most of the notification logic is abstracted to the level of NotifyNewOrder, NotifyDelegation, and NotifyOrderFulfilled - these should support simply doing nothing if Hipchat isn't configured.

File: core/core.go [hunk 3]
```diff
@@ -658,3 +728,128 @@ func Export(jsonIn []byte) ([]byte, error) {
...
+	}
+
+	if err := validateUser(o.Name, o.Password, false); err != nil {
+		return jsonStatusError(err)
+	}
+
+	// Get the owners of the ciphertext.
+	owners, _, err := crypt.GetOwners(o.Data)
+	if err != nil {
+		errors.New("Unable to find the ciphertext's owners.")
```
Comment (j-delaney) [inline core/core.go]: Should this be `err = errors.New`?

File: redoctober.go [hunk 4]
```diff
@@ -445,16 +453,18 @@ var indexHtml = []byte(`<!DOCTYPE html>
 					<div class="form-group row">
 						<div class="col-md-6">
 							<label for="user-name">User name</label>
-							<input type="text" name="Name" class="form-control" id="user-name" placeholder="User name" required />
+							<input type="text" name="Name" class="form-control" id="user-name" placeholder="User name"/ required>
```
Comment (j-delaney) [inline redoctober.go]: Why is the `/` after `placeholder="User name"` and not before the closing `>`?

File: passvault/passvault.go [hunk 5]
```diff
@@ -364,8 +365,16 @@ func (records *Records) AddNewRecord(name, password string, admin bool, userType
 }
 
 // ChangePassword changes the password for a given user.
-func (records *Records) ChangePassword(name, password, newPassword string) (err error) {
+func (records *Records) ChangePassword(name, password, newPassword, hipchatName string) (err error) {
 	pr, ok := records.GetRecord(name)
+
+	if len(hipchatName) != 0 {
```
Comment (j-delaney) [inline passvault/passvault.go]: It might make sense to rename this API endpoint if it's going to be expanding beyond just changing the password

File: redoctober.go [hunk 6]
```diff
@@ -582,6 +592,92 @@ var indexHtml = []byte(`<!DOCTYPE html>
...
 		</section>
+		<hr />
+		<section class="row">
+			<div id="orders" class="col-md-6">
+				<h3>Create Order</h3>
+
+				<form id="order" class="ro-user-order" role="form" action="/order" method="post">
+					<div class="feedback order-feedback"></div>
+					<div class="form-group">
+					<div class="form-group row">
```
Comment (j-delaney) [inline redoctober.go]: I think there might be a duplicate here. There is a `.form-group` within a `.form-group`

File: redoctober.go [hunk 7]
```diff
@@ -582,6 +592,92 @@ var indexHtml = []byte(`<!DOCTYPE html>
...
+							<label for="order-label">Label</label>
+							<input type="text" name="Label" class="form-control" id="order-user-label" placeholder="Label" required />
+						</div>
+						<div class="col-md-6">
+							<label for="order-duration">Duration</label>
+							<input type="text" name="Duration" class="form-control" id="order-duration" placeholder="Duration (e.g., 2h34m)" required />
+						</div>
+						<div class="col-md-6">
+							<label for="order-duration">Uses</label>
+							<input type="text" name="Uses" class="form-control" id="order-duration" placeholder="Uses" required />
```
Comment (j-delaney) [inline redoctober.go]: This `id` should probably be `order-uses` instead

File: redoctober.go [hunk 8]
```diff
@@ -805,6 +900,99 @@ var indexHtml = []byte(`<!DOCTYPE html>
...
+			$('body').on('submit', 'form#order', function(evt){
+				evt.preventDefault();
+				var $form = $(evt.currentTarget),
+					data = serialize($form);
+
+				submit( $form, {
+					data : data,
+					success : function(d){
+					d = JSON.parse(window.atob(d.Response));
+					$form.find('.feedback').empty().append(
```
Comment (j-delaney) [inline redoctober.go]: Instead of doing `.empty()` then `.append()` you could just do `$form.find('.feedback').html(...`

File: redoctober.go [hunk 9]
```diff
@@ -805,6 +900,99 @@ var indexHtml = []byte(`<!DOCTYPE html>
...
+
+				submit( $form, {
+					data : data,
+					success : function(d){
+					d = JSON.parse(window.atob(d.Response));
+					ordout = "";
+					for (var jj in d){
+						if (!d.hasOwnProperty(jj))
+							continue;
+						var o = d[jj]
```
Comment (j-delaney) [inline redoctober.go]: [nit] Missing semicolon

File: redoctober.go (hunk 8 above)
Comment (ejcx) [inline redoctober.go]: That's the pattern that was already here so I'm rolling with empty().append(). html is actually not safe to use in this situation either. XSS'able

Comment (j-delaney) [inline redoctober.go]: I understand keeping `.empty().append()` for consistency sake, especially because then it have the same structure as forms like `create-user` where it just does a `.append()`.

As for the security, `.html(value)` actually just ends up calling `.empty().append(value)` (see [source code](https://github.com/jquery/jquery/blob/22449eb/src/manipulation.js#L425)) so I believe they're equally susceptible to XSS

Comment (ejcx) [inline redoctober.go]: Oh that's super neat. Thanks for pointing that out.

Comment (ejcx) [inline redoctober.go]: Looks like red october is super XSSable

Comment (j-delaney) [inline redoctober.go]: I'd be happy to write a patch that finds all instances of `.html` being called (directly or indirectly) and using `.text` instead

Comment (ejcx) [inline redoctober.go]: Thanks but I've got it. Going to add it into this PR

**Comment by ejcx** [comment]
@j-delaney and @andrewbuss thanks for the review. I hadn't given it a review myself before I submitted the PR (to get us talking internally about if it was actually working how we wanted). It was very helpful

**Comment by j-delaney** [comment]
No problem! Glad we could help

File: core/core.go [hunk 10]
```diff
@@ -146,6 +173,26 @@ func jsonResponse(resp []byte) ([]byte, error) {
 	return json.Marshal(ResponseData{Status: "ok", Response: resp})
 }
 
+func getAltNameFromName(alt, name string) (altName string, found bool) {
```
Comment (Bren2010) [inline core/core.go]: Could these two functions be methods on passvault.Records?

File: order/order.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
+package order
+
+import (
+	//	"fmt"
```
Comment (Bren2010) [inline order/order.go]: Left-over?

File: order/order.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+type Order struct {
+	Name string
+	Num  string
+
+	TimeRequested     time.Time
+	ExpiryTime        time.Time
+	DurationRequested time.Duration
+	Delegated         int
+	ToDelegate        int
```
Comment (Bren2010) [inline order/order.go]: ToDelegate isn't used anywhere else?

File: keycache/keycache.go [hunk 13]
```diff
@@ -305,3 +306,33 @@ func (cache *Cache) DecryptShares(in [][]byte, name, user string, labels []strin
 
 	return
 }
+
+func (cache *Cache) DelegateStatus(name string, label string, admins []string) (adminsDelegated []string, hasDelegated int) {
```
Comment (Bren2010) [inline keycache/keycache.go]: This needs function documentation.

File: core/core.go [hunk 14]
```diff
@@ -183,23 +229,37 @@ func validateName(name, password string) error {
...
+		orders.Hipchat = hipchat.HipchatClient{
+			ApiKey: hcKey,
+			RoomId: roomId,
+			HcHost: hcHost,
+			RoHost: roHost,
+		}
+	}
 	cache = keycache.Cache{UserKeys: make(map[keycache.DelegateIndex]keycache.ActiveUser)}
-	crypt = cryptor.New(&records, &cache)
+	orders.PrepareOrders()
```
Comment (Bren2010) [inline core/core.go]: You could create an `orders.NewOrderer(hcKey, ..., hcHost) orders.Orderer` function and move the PrepareOrders and HipChat logic above into it? The NewXXX pattern is what I've usually seen for things that need initialization.

File: index.html [hunk 15]
```diff
@@ -309,6 +311,93 @@ <h1 class="page-header">Red October Management</h1>
 				</form>
 			</div>
 		</section>
+		<hr />
+		<section class="row">
+			<div id="orders" class="col-md-6">
+				<h3>Create Order</h3>
```
Comment (Bren2010) [inline index.html]: There's no link in the navbar for this. This entire file also needs to be synced with redoctober.go.

File: cryptor/cryptor.go [hunk 16]
```diff
@@ -29,10 +30,11 @@ const (
 type Cryptor struct {
 	records *passvault.Records
 	cache   *keycache.Cache
+	orders  *order.Orderer
```
Comment (Bren2010) [inline cryptor/cryptor.go]: I don't see that this is ever used?

Comment (ejcx) [inline cryptor/cryptor.go]: Nice catch here.

File: core/core.go (hunk 14 above)
Comment (ejcx) [inline core/core.go]: Also agreed

File: order/order.go (hunk 11 above)
Comment (ejcx) [inline order/order.go]: my auto-gofmt is slacking here!

Comment (grittygrease) [inline order/order.go]: goimports should help with this sort of thing

File: order/order.go [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+	OrderFor string
+
+	OrderId     string
+	OrderOwners []string
+}
+
+// Orders represents a mapping of Order IDs to Orders. This structure
+// is useful for looking up information about individual Orders and
+// whether or not an order has been fulfilled. Orders that have been
+// fulfilled will removed from the structure.
```
Comment (kisom) [inline order/order.go]: s/will removed/will be removed/

**Comment by kisom** [comment]
Go side LGTM.

File: core/core.go [hunk 18]
```diff
@@ -183,21 +215,35 @@ func validateName(name, password string) error {
 }
 
 // Init reads the records from disk from a given path
-func Init(path string) error {
+func Init(ca, hcKey, hcRoom, hcHost, roHost string) error {
```
Comment (grittygrease) [inline core/core.go]: Why was this changed to "ca" from "path"? It still represents the location of the vault on disk.

Comment (ejcx) [inline core/core.go]: That would be a mistake of mine.  Will fix

File: core/core.go [hunk 19]
```diff
@@ -106,6 +112,33 @@ type ExportRequest struct {
 	Password string
 }
 
+type OrderRequest struct {
+	Name     string
+	Password string
+	Duration string
+	Uses     string
```
Comment (jacobhaven) [inline core/core.go]: This should be an `int`, to match the other Request structs if nothing else.

File: core/core.go [hunk 20]
```diff
@@ -106,6 +112,33 @@ type ExportRequest struct {
...
+	Data     []byte
+	Label    string
```
Comment (jacobhaven) [inline core/core.go]: Can we make this `Labels []string`?

File: core/core.go [hunk 21]
```diff
@@ -661,3 +730,158 @@ func Export(jsonIn []byte) ([]byte, error) {
...
+	if ord, ok := orders.Orders[o.OrderNum]; ok {
+		if out, err = json.Marshal(ord); err != nil {
+			return jsonStatusError(err)
+		} else if len(out) == 0 {
+			return jsonStatusError(errors.New("No order with that number"))
+		}
+
+		return jsonResponse(out)
+	}
+	return
```
Comment (jacobhaven) [inline core/core.go]: This should be restructured so that `jsonStatusError(errors.New("No order with that number"))` is returned if `orders.Orders` doesn't contain `o.OrderNum`

File: order/order.go [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+)
+
+const (
+	NewOrder       = "%s has created an order for the label %s. requesting %s delegations for %s"
+	NewOrderLink   = "@%s - https://%s?%s"
+	OrderFulfilled = "%s has had order %s fulfilled."
+	NewDelegation  = "%s has delegated the label %s to %s (per order %s) for %s"
+)
+
+type Order struct {
```
Comment (jacobhaven) [inline order/order.go]: This is marshalled to JSON and returned by the API, it should probably have the `json` tags for doing that.

File: order/order.go [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+	Name string
+	Num  string
+
+	TimeRequested     time.Time
+	ExpiryTime        time.Time
```
Comment (jacobhaven) [inline order/order.go]: `ExpiryTime` doesn't do anything and should probably be removed.

File: order/order.go [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+
+const (
+	NewOrder       = "%s has created an order for the label %s. requesting %s delegations for %s"
+	NewOrderLink   = "@%s - https://%s?%s"
+	OrderFulfilled = "%s has had order %s fulfilled."
+	NewDelegation  = "%s has delegated the label %s to %s (per order %s) for %s"
+)
+
+type Order struct {
+	Name string
```
Comment (jacobhaven) [inline order/order.go]: `Name` should probably be renamed to `Owner` (i.e. the user who placed the order request)
and (as @jkroll-cf suggested) a `Users  []string` field for the users that delegations are requested for (possibly empty)
`Label` should also instead be `Labels []string`

**Comment by grittygrease** [comment]
Comments sent out of band.

File: core/core.go (hunk 18 above)
Comment (grittygrease) [inline core/core.go]: Can you update this?

Comment (ejcx) [inline core/core.go]: (boom)

File: README.md [hunk 25]
```diff
@@ -247,6 +247,78 @@ Example input JSON format:
            -d '{"Name":"Alice","Password":"Lewis"}'
     {"Status":"ok"}
 
+
+### Order
+
+Order creates a new order and lets other users know delegations are needed.
+
+Example input JSON format:
```
Comment (grittygrease) [inline README.md]: Can you make the spacing consistent in these JSON examples?

File: cmd/ro/main.go [hunk 26]
```diff
@@ -99,7 +105,7 @@ func runDelegate() {
 		Name:     user,
 		Password: pswd,
 		Uses:     uses,
-		Time:     time,
+		Time:     duration,
```
Comment (grittygrease) [inline cmd/ro/main.go]: nice change

File: core/core.go (hunk 1 above)
Comment (grittygrease) [inline core/core.go]: This achieves the current goal, but eventually we should move to a more generalized notification system.

File: core/core.go (hunk 21 above)
Comment (grittygrease) [inline core/core.go]: @ejcx can you update the code to pick up this change?

File: order/order.go (hunk 22 above)
Comment (grittygrease) [inline order/order.go]: I believe that @ejcx is just following the convention of the project. Let's clean up the tagging in a later PR.

File: core/core.go (hunk 21 above)
Comment (ejcx) [inline core/core.go]: nice catch jacob.

File: cmd/ro/main.go [hunk 27]
```diff
@@ -45,7 +50,7 @@ func registerFlags() {
 	flag.StringVar(&owners, "owners", "", "comma separated owner list")
 	flag.StringVar(&users, "users", "", "comma separated user list")
 	flag.IntVar(&uses, "uses", 0, "number of delegated key uses")
-	flag.StringVar(&time, "time", "0h", "duration of delegated key uses")
+	flag.StringVar(&duration, "time", "0h", "duration of delegated key uses")
```
Comment (lziest) [inline cmd/ro/main.go]: flag.DurationVar is better.

Comment (lziest) [inline cmd/ro/main.go]: nvm, I see the existing code.

**Comment by jacobhaven** [comment]
@ejcx Could you review my [client code](https://github.com/cloudflare/redoctober/pull/106/files#diff-1cd41a2611da49ce63630d5a51bd82c5R285) and check that it works with your more recent changes?

File: cmd/ro/main.go [hunk 28]
```diff
@@ -207,6 +213,33 @@ func runDecrypt() {
 	ioutil.WriteFile(outPath, msg.Data, 0644)
 }
 
+func runOrder() {
+	req := core.OrderRequest{
+		Name:     user,
+		Password: pswd,
+		Uses:     uses,
+		Duration: duration,
```
Comment (jacobhaven) [inline cmd/ro/main.go]: `Users:    processCSL(users)` Should be added here
